Type changing lenses compatible with nominal roles

Lysxia opened the pull request:
Closes #55.

Lysxia commented on src/Data/Generics/Product/Fields.hs:
```diff
@@ -159,6 +168,18 @@ instance  -- see Note [Changing type parameters]
+  , Generic t
+  , GHasKey  field (Rep s) (Rep t) a b
+  , UnifyHead s t
+  ) => HasField_ field s t a b where
```
Should I keep the `GHasKey' field (Rep s) a` from the `HasField` instance? What do those `GHasKey'` constraints do?

Lysxia commented:
That sounds like a good plan to me!

Lysxia commented:
Actually, were you saying this PR is a breaking change, because it's only defining a new class (see my latest comments in #55).

kcsongor commented:
Right, sorry - I was thinking that it might make sense to lose those fundeps, as they seem to be a small price to pay for the simplicity of `UnifyHead`. I think that in practice type inference shouldn't suffer because that instance is general enough to be picked unless there's an overlapping one.

Lysxia commented:
That makes sense!

kcsongor commented:
Sorry I took quite a while to look at this again! Unfortunately it does seem like the extra functional dependencies are required when composing polymorphic lenses, such as in this scenario: https://github.com/kcsongor/generic-lens/issues/62
Incidentally, the above broke because I deleted the `GHasKey' field (Rep s) a` constraint from the lens (thinking it was redundant), but now we know why it was needed!
I tried the above with `field_`, but the intermediate types in the composition can't be resolved due to the missing fundeps.

Quite unfortunate, because this means we need two different ways for doing very similar things (at least for now). Do you have any ideas on how to best incorporate this into the API? In particular, I think the `UnifyHead` version should be used in most of the cases, and the indexing one when you need a bit better type inference.

Lysxia commented:
Changing one `UnifyHead` instance from

```haskell
instance UnifyHead f g => UnifyHead (f a) (g b)
```

to

```haskell
instance (UnifyHead f g, gb ~ g b) => UnifyHead (f a) gb
```

and adding both constraints to `HasField_`, etc.:

```haskell
( UnifyHead s t
, UnifyHead t s
, ... )
```

seems to pass the test #62.

I also added `HasField0`, and both `HasField` and `HasField_` are now implemented using `HasField0`.

kcsongor commented on src/Data/Generics/Product/Positions.hs:
```diff
@@ -147,6 +165,37 @@ coerced = coerce
+  , HasTotalPositionP i (CRep s) ~~ 'Just a
+  , HasTotalPositionP i (CRep t) ~~ 'Just b
+  , HasTotalPositionP i (CRep (Indexed s)) ~~ 'Just a'
+  , HasTotalPositionP i (CRep (Indexed t)) ~~ 'Just b'
```
Should these `Indexed` ones be removed from here? I think they're redundant (same for `HasPosition_`), given that `Infer` is no longer using them.

Lysxia replied:
Are you proposing to remove all `HasTotal*** ... (CRep (Indexed _)) ...`?

kcsongor commented:
This looks very good, thanks for adding the documentation too! I'll find some time tomorrow to merge it in and make a release with the new changes. Thanks!
